src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -0,0 +1,238 @@
+                [_storageDelegate controller:_controller
+                           removeValueForKey:ResumptionByNodeIDKey(oldInfo.nodeID)
+                               securityLevel:MTRStorageSecurityLevelSecure
+                                 sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+            });
+        }
+    }
+
+    [_nodesWithResumptionInfo removeAllObjects];
+}
+
+- (CHIP_ERROR)storeLastUsedNOC:(MTRCertificateTLVBytes)noc
+{
+    __block BOOL ok;
+    dispatch_sync(_storageDelegateQueue, ^{
+        ok = [_storageDelegate controller:_controller
+                               storeValue:noc
+                                   forKey:sLastUsedNOCKey
+                            securityLevel:MTRStorageSecurityLevelSecure
+                              sharingType:MTRStorageSharingTypeSameIdentityRequired];
+    });
+    return ok ? CHIP_NO_ERROR : CHIP_ERROR_PERSISTED_STORAGE_FAILED;
+}
+
+- (MTRCertificateTLVBytes _Nullable)fetchLastUsedNOC
+{
+    __block id data;
+    dispatch_sync(_storageDelegateQueue, ^{
+        data = [_storageDelegate controller:_controller
+                                valueForKey:sLastUsedNOCKey
+                              securityLevel:MTRStorageSecurityLevelSecure
+                                sharingType:MTRStorageSharingTypeSameIdentityRequired];
+    });
+
+    if (data == nil) {
+        return nil;
+    }
+
+    if (![data isKindOfClass:[NSData class]]) {
+        return nil;
+    }
+
+    return data;
+}
+
+- (nullable MTRCASESessionResumptionInfo *)_findResumptionInfoWithKey:(NSString *)key
+{
+    __block id resumptionInfo;
+    dispatch_sync(_storageDelegateQueue, ^{
+        resumptionInfo = [_storageDelegate controller:_controller
+                                          valueForKey:key
+                                        securityLevel:MTRStorageSecurityLevelSecure
+                                          sharingType:MTRStorageSharingTypeSameIdentityAllowed];
+    });
+
+    if (resumptionInfo == nil) {
+        return nil;
+    }
+
+    if (![resumptionInfo isKindOfClass:[MTRCASESessionResumptionInfo class]]) {
+        return nil;
+    }
+
+    return resumptionInfo;
+}
+
+@end
+
+@implementation MTRCASESessionResumptionInfo
+
+#pragma mark - NSSecureCoding
+
+static NSString * const sNodeIDKey = @"nodeID";
+static NSString * const sResumptionIDKey = @"resumptionID";
+static NSString * const sSharedSecretKey = @"sharedSecret";
+static NSString * const sCATsKey = @"CATs";
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+- (instancetype _Nullable)initWithCoder:(NSCoder *)decoder
+{
+    _nodeID = [decoder decodeObjectOfClass:[NSNumber class] forKey:sNodeIDKey];
+    _resumptionID = [decoder decodeObjectOfClass:[NSData class] forKey:sResumptionIDKey];
+    _sharedSecret = [decoder decodeObjectOfClass:[NSData class] forKey:sSharedSecretKey];
+    _caseAuthenticatedTags = [decoder decodeObjectOfClass:[NSSet class] forKey:sCATsKey];
+    return self;
```

Comment:
Do we need some internal sanity testing here, and throw if not decoded properly?

Reply:
My understanding is that decodeObjectOfClass will throw already if the decode fails on that level, so we know we actually have an NSNumber, two NSData, and an NSSet, right?

1. We could check for the NSNumber being an integer > 0, but I am not sure how to actually check the "an integer" condition for an NSNumber.  `objCType` seems extremely hacky....
2. We could check that resumptionID has the right length (16) and sharedSecret has the right length (at most 32, but this may be subject to change as Matter evolves).  But this is already checked by the consumer of this API anyway.
3. We could check that the NSSet contains numbers and that they are valid CAT values.  This last part is checked by the consumer of this API, but the "contains numbers" bit is in fact a missing check that we should do here.

Thoughts?

Reply:
@ksperling-apple suggested encoding the tags as an array and decoding with decodeArrayOfObjectsOfClass which should ensure we have the right things.

Reply:
We do care, and it's not about bad clients, it's just data corruption protection, and bugs :)

Reply:
Unresolving conversation - if we care, then we should do some checking here, given the unarchiver doesn't check default-allowed classes - in other words, if somehow there's data corruption and we get an NSString instead of an NSData, then it could crash when we ask it for bytes.

Reply:
https://github.com/project-chip/connectedhomeip/pull/28533/commits/2ab709da18809e3fc2905119b03b6226aeb0b7fd should fix that.